pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,284 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the adjoint_metric_tensor.
+"""
+import warnings
+from pennylane import numpy as np
+
+import pennylane as qml
+
+
+def _get_generator(op):
+    """Reads out the generator and prefactor of an operation and converts
+    to matrix if necessary.
+
+    Args:
+        op (:class:`~.Operation`): Operation to obtain the generator of.
+    Returns:
+        array[float]: Generator matrix
+        float: Prefactor of the generator
+    """
+
+    generator, prefactor = op.generator
+    if not isinstance(generator, np.ndarray):
+        generator = generator.matrix
+    if op.inverse:
+        generator = generator.conj().T
+        prefactor *= -1
+
+    return generator, prefactor
+
+
+def _apply_any_operation(state, op, device, invert=False):
+    """Wrapper that allows to apply a variety of operations---or groups
+    of operations---to a state or to prepare a new state.
+    If ``invert=True``, this function makes sure not to alter the operations.
+    The state of the device, however may be altered, depending on the
+    device and performed operation(s).
+    """
+    # pylint: disable=protected-access
+    if isinstance(op, list):
+        if invert:
+            op = op[::-1]
+        for _op in op:
+            state = _apply_any_operation(state, _op, device, invert)
+        return state
+    if isinstance(op, qml.QubitStateVector):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_state_vector(op.parameters[0], op.wires)
+        return device._state
+    if isinstance(op, qml.BasisState):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_basis_state(op.parameters[0], op.wires)
+        return device._state
+
+    if invert:
+        op.inv()
+    state = device._apply_operation(state, op)
```

Comment:
from my reading of this function, it is actually _only_ limited to working with `default.qubit`, since no other devices are guaranteed to have these private methods. Should we mention this in the `adjoint_metric_tensor` docstring?

Reply:
Oh, that is a very good point. Yes, we should mention it, I am not sure how easy it would be to extend it to other devices?